### PR TITLE
Let RESTEasy Reactive exception mappers handle authentication failure with disabled proactive security

### LINKED ISSUE #22971: Authentication: cannot map `AuthenticationFailedException`
### Describe the bug

`HttpSecurityRecorder` uses this exception mapper:

```java
                //register the default auth failure handler
                //if proactive auth is used this is the only one
                //if using lazy auth this can be modified downstream, to control authentication behaviour
                event.put(QuarkusHttpUser.AUTH_FAILURE_HANDLER, new BiConsumer<RoutingContext, Throwable>() {
                    @Override
                    public void accept(RoutingContext routingContext, Throwable throwable) {
                        throwable = extractRootCause(throwable);
                        //auth failed
                        if (throwable instanceof AuthenticationFailedException) {
                            authenticator.sendChallenge(event).subscribe().with(new Consumer<Boolean>() {
                                @Override
                                public void accept(Boolean aBoolean) {
                                    if (!event.response().ended()) {
                                        event.response().end();
                                    }
                                }
                            }, new Consumer<Throwable>() {
                                @Override
                                public void accept(Throwable throwable) {
                                    event.fail(throwable);
                                }
                            });
                        } else if (throwable instanceof AuthenticationCompletionException) {
                            event.response().setStatusCode(401);
                            event.response().end();
                        } else if (throwable instanceof AuthenticationRedirectException) {
                            AuthenticationRedirectException redirectEx = (AuthenticationRedirectException) throwable;
                            event.response().setStatusCode(redirectEx.getCode());
                            event.response().headers().set(HttpHeaders.LOCATION, redirectEx.getRedirectUri());
                            event.response().headers().set(HttpHeaders.CACHE_CONTROL, "no-store");
                            event.response().headers().set("Pragma", "no-cache");
                            event.response().end();
                        } else {
                            event.fail(throwable);
                        }
                    }
```

Which means that some exceptions are handled directly, and can't be mapped by RESTEasy Reactive, even though when using non-proactive authentication, RR still comes later into play and will try to map the exception and write it out and will cause an ignored exception because the response has already been written.

Note that `authenticator.sendChallenge(event)` also sets the response in `HttpAuthenticator`:

```java
    public Uni<Boolean> sendChallenge(RoutingContext routingContext) {
        //we want to consume any body content if present
        //challenges won't read the body, and if we don't consume
        //things can get stuck
        routingContext.request().resume();
        Uni<Boolean> result = null;

        HttpAuthenticationMechanism matchingMech = routingContext.get(HttpAuthenticationMechanism.class.getName());
        if (matchingMech != null) {
            result = matchingMech.sendChallenge(routingContext);
        }
        if (result == null) {
            result = mechanisms[0].sendChallenge(routingContext);
            for (int i = 1; i < mechanisms.length; ++i) {
                HttpAuthenticationMechanism mech = mechanisms[i];
                result = result.onItem().transformToUni(new Function<Boolean, Uni<? extends Boolean>>() {
                    @Override
                    public Uni<? extends Boolean> apply(Boolean authDone) {
                        if (authDone) {
                            return Uni.createFrom().item(authDone);
                        }
                        return mech.sendChallenge(routingContext);
                    }
                });
            }
        }
        return result.onItem().transformToUni(new Function<Boolean, Uni<? extends Boolean>>() {
            @Override
            public Uni<? extends Boolean> apply(Boolean authDone) {
                if (!authDone) {
                    routingContext.response().setStatusCode(401);
                    routingContext.response().end();
                }
                return Uni.createFrom().item(authDone);
            }
        });
    }
```

### Expected behavior

I want to be able to map the exceptions to what I want, without having to register a Vert.x route just to be able to override the `QuarkusHttpUser.AUTH_FAILURE_HANDLER`, and I'd prefer if it were mapped by RESTEasy Reactive using the existing exception mapping process.

CC @stuartwdouglas @sberyozkin @geoand 

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### LINKED ISSUE #25732: Unable to change unauthenticated response with `quarkus-resteasy-reactive`
### Describe the bug

I tried to add a custom exception mapper for unauthenticated exceptions. Followed the instruction on https://quarkus.io/guides/security-built-in-authentication#how-to-customize-authentication-exception-responses. However, although the exception mapper is defined, it's never being invoked. Switching the proactive auth on and off does not have any impact.

The exception mapper is the same as in the documentation:

```
import io.quarkus.security.AuthenticationFailedException;

import javax.annotation.Priority;
import javax.ws.rs.Priorities;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.UriInfo;
import javax.ws.rs.ext.ExceptionMapper;
import javax.ws.rs.ext.Provider;


@Provider
@Priority(Priorities.AUTHENTICATION)
public class AuthenticationFailedExceptionMapper implements ExceptionMapper<AuthenticationFailedException> {

    @Context
    UriInfo uriInfo;

    @Override
    public Response toResponse(AuthenticationFailedException exception) {
        // on purpose change to 5xx
        return Response.status(500).entity("Custom body").build();
    }
}
```

### Expected behavior

Response can be customized.

### Actual behavior

Always returns the same response:

```
$ curl 'http://localhost:8080/hello' -u scott:wrong -v
*   Trying 127.0.0.1:8080...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8080 (#0)
* Server auth using Basic with user 'scott'
> GET /hello HTTP/1.1
> Host: localhost:8080
> Authorization: Basic c2NvdHQ6cmVhZGU=
> User-Agent: curl/7.68.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
* Authentication problem. Ignoring this.
< www-authenticate: basic realm="Quarkus"
< content-length: 0
< 
* Connection #0 to host localhost left intact
```

### How to Reproduce?

Here is the reproducible example: https://github.com/ivansenic/quarkus-25732

### Output of `uname -a` or `ver`

Linux ise-Precision-5550 5.13.0-41-generic #46~20.04.1-Ubuntu SMP Wed Apr 20 13:16:21 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.3" 2022-04-19 OpenJDK Runtime Environment (build 17.0.3+7-Ubuntu-0ubuntu0.20.04.1) OpenJDK 64-Bit Server VM (build 17.0.3+7-Ubuntu-0ubuntu0.20.04.1, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.9.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven home: /home/ise/.m2/wrapper/dists/apache-maven-3.8.4-bin/52ccbt68d252mdldqsfsn03jlf/apache-maven-3.8.4 Java version: 17.0.3, vendor: Private Build, runtime: /usr/lib/jvm/java-17-openjdk-amd64 Default locale: en_US, platform encoding: UTF-8 OS name: "linux", version: "5.13.0-41-generic", arch: "amd64", family: "unix"

### Additional information

_No response_

### LINKED ISSUE #26314: @Provider unable to handle Exceptions / default Quarkus error response is used instead
### Describe the bug

Since we are handling all Exceptions with the  `ExceptionMapper<Throwable>` we should never get the default Quarkus error response. 
There is at least one case that an Exception is not handled by our Custom ExceptionMapper.

### Expected behavior

Error response when we call http://localhost:8080/hello should always be:

 "This is custom handler!"

### Actual behavior

Error response when Keycloak server is offline and we call http://localhost:8080/hello with Authorization header in place

```
{
    "details": "Error id b849414c-7207-4533-9cfd-4206733da866-1, io.quarkus.oidc.OIDCException: OIDC server is not available at the 'http://localhost:8001/realms/testing' URL. Please make sure it is correct. Note it has to end with a realm value if you work with Keycloak, for example: 'https://localhost:8180/auth/realms/quarkus'",
    "stack": "io.quarkus.oidc.OIDCException: OIDC server is not available at the 'http://localhost:8001/realms/testing' URL. Please make sure it is correct. Note it has to end with a realm value if you work with Keycloak, for example: 'https://localhost:8180/auth/realms/quarkus'\r\n\tat io.quarkus.oidc.runtime.OidcRecorder.toOidcException(OidcRecorder.java:236)\r\n\tat io.quarkus.oidc.runtime.OidcRecorder$5.apply(OidcRecorder.java:305)\r\n\tat io.quarkus.oidc.runtime.OidcRecorder$5.apply(OidcRecorder.java:299)\r\n\tat io.smallrye.context.impl.wrappers.Slo
```

image

### How to Reproduce?

1. Generate a new Quarkus project:
image

2. Change GreetingResource.java to:
``` java
package org.acme;

import io.quarkus.arc.Priority;

import javax.annotation.security.RolesAllowed;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.ext.ExceptionMapper;
import javax.ws.rs.ext.Provider;

@Path("/hello")
public class GreetingResource {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    @RolesAllowed("user")
    public String hello() {
        throw new IllegalStateException("Forcing error that handle successfully");
        //return "Hello from RESTEasy Reactive";
    }

    @Provider
    @Priority(Integer.MIN_VALUE)
    public static class AllMissingExceptionsHandler implements ExceptionMapper<Throwable> {

        @Override
        public Response toResponse(Throwable throwable) {
            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity("This is custom handler!").build();
        }
    }

}
```

3. Change application.properties to:
```
quarkus.oidc.client-id=testing-service
quarkus.oidc.auth-server-url=http://localhost:8001/realms/testing
quarkus.keycloak.devservices.enabled=false
```
4. Call http://localhost:8080/hello with postman by providing a new header 
`Authorization:Bearer Just_a_Random_Text` (image)

image

5. Response is not the expected one (expected:  "This is custom handler!")

### Output of `uname -a` or `ver`

MINGW64_NT-10.0-22000 spirospc 3.3.3-341.x86_64 2022-01-17 11:45 UTC x86_64 Msys

### Output of `java -version`

openjdk version "17.0.2" 2022-01-18 OpenJDK Runtime Environment (build 17.0.2+8-86) OpenJDK 64-Bit Server VM (build 17.0.2+8-86, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.10.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.5 (3599d3414f046de2324203b78ddcf9b5e4388aa0) Maven home: C:\tools\apache-maven-3.8.5-bin\apache-maven-3.8.5 Java version: 17.0.2, vendor: Oracle Corporation, runtime: C:\tools\openjdk-17.0.2\jdk-17.0.2 Default locale: en_US, platform encoding: Cp1252 OS name: "windows 11", version: "10.0", arch: "amd64", family: "windows"

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by michalvavrik**
fixes: #25732
fixes:  #22971
fixes: #26314 (for proactive=false)

There are situations (e.g. permissions check, see #22971) when authentication is performed before RESTEasy Reactive begin processing, thus simply removing default authentication failure handler won't do. Users expects that exceptions mappers are going to be invoked (which is not a case).

This PR provides it's own auth failure handler that fails the event and let RR abort handlers to deal with exceptions. It intentionally handles only `proactive=false` scenarios (still, it fixes linked issues). It's necessary to differ between situations when RR begun procession (then we shouldn't fail event as it would be handled by both failure handlers and abort handlers) and didn't begin processing. I added `ExceptionMapperTest`, IT version can't be added as it requires disabled proactive security (build time property, so test profile won't do), we don't have suitable module for this.

### CODE REVIEW

**Comment by sberyozkin** [comment]
Hi @michalvavrik Can you please add a test alongside the one you already added, which will use Form Auth and without a custom exception mapper, just to verify we'll get 302 in this case.
Thanks

**Comment by michalvavrik** [comment]
Hello @sberyozkin, I added the simplest test I could find. If you want any adjustments/more tests, no problem, just specify it in detail, thank you.

**Comment by sberyozkin** [comment]
@michalvavrik Thanks, it is good to have a test confirming that the default handler is operational without the mapper. LGTM, lets wait till tomorrow, in case Stuart or Georgios will have something to add

**Review by sberyozkin — APPROVED** [review]
Thanks @michalvavrik, resolving 3 issues with one PR is great :-)

**Comment by michalvavrik** [comment]
I checked `FlywayExtensionInitSqlTest` and it don't seem to be related. Thank you Sergey. Sure, let's wait for review from RR guys however long you prefer, no rush.

**Comment by sberyozkin** [comment]
Lets merge it now since no more comments have followed, we can tweak things as always going forward :-)

**Pull request merged**
